packages/elements/src/elementSlice.js

```diff
@@ -0,0 +1,28 @@
+/*
+ * Copyright 2022 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+const elementSlice = (set) => ({
+  state: {
+    elementTypes: {},
+  },
+  actions: {
+    registerElementTypes: (elementType) =>
+      set((data) => {
+        data.element.state.elementTypes[elementType.type] = elementType;
+      }),
```

Comment:
Is there a reason for directly mutating the store vs returning a new `element` property?

Reply:
if this is `store.actions.registerElementTypes(type)` should we either accept multiple types? or make this `store.actions.registerElementType` ?

Reply:
maybe since we got the whole slice abstraction goin, might make it nicer to refer to slices vs data:
```js
registerElementTypes: (elementTypes) =>
      set((slices) => ({
        element: {
          ...slices.element,
          state: {
            ...slices.element.state,
            elementTypes: [ ...slices.element.state.elementTypes, ...elementTypes ],
          }
        }
      })),
```

The separation of `state` & `actions` here makes it a little verbose. Not sure what popular consensus is, but I'd be in favor of a flat structure vs separating each slice into `state` & `actions`. 

Honestly, have never been the biggest fan of the split between `state` & `actions` in any of our code. Seems like it's always been a pure stylistic choice that just adds another layer to keeping track of object references

Reply:
> Is there a reason for directly mutating the store vs returning a new element property?

Yes, a new element should have been returned, updated in https://github.com/GoogleForCreators/web-stories-wp/pull/10964/commits/a2691d7039061cb42e09135421de58972ab3f031. 

> if this is store.actions.registerElementTypes(type) should we either accept multiple types? or make this store.actions.registerElementType ?

I think either is fine. We already have `registerElementTypes` utility function which does that for multiple element types.

> The separation of state & actions here makes it a little verbose. Not sure what popular consensus is, but I'd be in favor of a flat structure vs separating each slice into state & actions.

> Honestly, have never been the biggest fan of the split between state & actions in any of our code. Seems like it's always been a pure stylistic choice that just adds another layer to keeping track of object references

I have a couple of reasons to keep separate `state` and `actions`

1. We are already keeping separate `state` and `actions` in core packages, so it will be consistent.
2. It is much easier to create API like `select( 'sliceName' )` and `dispatch('sliceName' ).someAction()` for retrieving and updating state like Gutenberg `wp.data.select` and `wp.data.dispatch`

WDYT?

Reply:
All this sounds great to me. Ah yes, wasn't considering external api. Thanks for elaborating!